proxies.py

```diff
@@ -0,0 +1,183 @@
+    """
+    Simple proxy that expects both sides to support websockets
+    """
+
+    def _recv_line(self, sock):
+        """
+        Receive from the socket on char at a time
+        until we get to a new line
+
+        Args:
+          sock: The socket to receive from
+        """
+        line = []
+        while True:
+            character = sock.recv(1)
+            line.append(character)
+            if character == six.b("\n"):
+                break
+        return six.b("").join(line)
+
+    def _read_headers(self, sock):
+        """
+        Read the headers line by line, we previously send to the target sock
+        
+        Args:
+          sock: The socket we have send headers to
+        """
+        status = None
+        headers = {}
+
+        while True:
+            line = self._recv_line(sock)
+            line = line.decode('utf-8').strip()
+            if not line:
+                break
+
+            if not status:
+                status_info = line.split(" ", 2)
+                status = int(status_info[1])
+
+            else:
+                kv = line.split(":", 1)
+                if len(kv) == 2:
+                    field, value = kv
+                    headers[field.lower()] = value.strip().lower()
+                else:
+                    raise Exception("Invalid header: %s" % kv)
+
+        return status, headers
+
+    def new_websocket_client(self):
+        """
+        Create a new socket to the target
+        """
+        parse = urlparse.urlparse(self.path)
+        args = urlparse.parse_qs(parse.query)
+        host = args.get("host", [""]).pop()
+        port = 443
+        ticket = args.get("ticket", [""]).pop()
+
+        #Consider using the websocket_client code
+        tsock = websockify.websocket.WebSocketServer.socket(
+            host, port, connect=True, use_ssl=True, unix_socket=False)
+        request = "GET /ticket/%s HTTP/1.1\r\n" % ticket
+        upgrade = request + str(self.headers) + "\r\n"
+        print upgrade
+        tsock.send(upgrade)
+        status, headers = self._read_headers(tsock)
+        if status != 101:
+            raise Exception("Bad Response, %s, %s" % (status, headers))
+        self.do_proxy(tsock)
+
+    def do_proxy(self, target):
+        """
+        Handle the reading and writing of data being passed through
+        the proxy
+
+        Args:
+          target: the websockify WebSocketServer socket of the target
+            server
+        """
+        cqueue = []
+        tqueue = []
+        rlist = [self.request, target]
+        self.request.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+        target.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+
+        if self.server.heartbeat:
+            now = time.time()
+            self.heartbeat = now + self.server.heartbeat
+        else:
+            self.heartbeat = None
+
+        while True:
+            wlist = []
+
+            if self.heartbeat is not None:
```

Comment:
Since the sending of the heartbeat ping was causing problems and was removed, is this heartbeat code even needed anymore?